packages/apps/patterns/react-ui/src/panels/JoinPanel/joinMachine.ts

```diff
@@ -79,39 +80,49 @@ const getInvitationSubscribable = (
       onError?: (error: any) => void,
       complete?: () => void
     ): Subscription => {
-      const unsubscribe = invitation.subscribe({
-        onAuthenticating: (invitation: Invitation) => {
-          log('[invitation authenticating]', { Domain, invitation });
-          return next({ type: `authenticate${Domain}Invitation`, invitation });
-        },
-        onCancelled: () => {
-          log.warn('[invitation cancelled]', { Domain });
-          return next({ type: `fail${Domain}Invitation`, reason: 'cancelled' } as FailInvitationEvent);
-        },
-        onTimeout: () => {
-          log.error('[invitation timeout]', { Domain });
-          return next({ type: `fail${Domain}Invitation`, reason: 'timeout' } as FailInvitationEvent);
-        },
-        onConnecting: (invitation: Invitation) => {
-          log('[invitation connecting]', { Domain, invitation });
-          return next({ type: `connect${Domain}Invitation`, invitation });
-        },
-        onConnected: (invitation: Invitation) => {
-          log('[invitation connected]', { Domain, invitation });
-          return next({ type: `connectionSuccess${Domain}Invitation`, invitation });
-        },
-        onSuccess: () => {
-          log('[invitation success]', { Domain });
-          next({ type: `succeed${Domain}Invitation` });
-          return complete?.();
+      const subscription = invitation.subscribe(
```

Comment:
> what if there's an event that isn't represented by a state change

I agree, observables should emit events, since the state of whatever they represent is (in theory) not relevant, and emitting states is lossy for this reason. For example, if the user submits the wrong authentication code, the `invitationObservable` will emit the same state as when it’s ready for the user to input the authentication code for the first time, so the UI state has to respond to that event differently depending on the UI’s current state, which is not the right source of truth.

Reply:
If we should also know the invitation’s state, we couldmake invitations into actual XState machines.